File: lab3/src/shapes/Trapezoid.java
package shapes;

import geometry.Line;
import geometry.Point;

public class Trapezoid extends Shape {

    protected Trapezoid(Point[] points) {
        super(points);
    }

    @Override
    public double getArea() {
        var edges = this.getEdgeLines();

        var a = edges[0].length();
        var b = edges[1].length();
        var c = edges[2].length();
        var d = edges[3].length();

        var T = a - c;
        var D = d * d - b * b;
        var R = (T * T - D) / (2 * T);
        var h = Math.sqrt(b * b - R * R);

        var area = h * (a + c) / 2;

        return area;
    }

    @Override
    public boolean isExists() {
        var edges = this.getEdgeLines();
        
        var bottom = edges[0];
        var top = edges[2];
        var left = edges[1];
        var right = edges[3];

        return Line.isParallel(bottom, top) && !Line.isParallel(left, right);
    }
}

File: lab5/src/shapes/Trapezoid.kt
package shapes

import geometry.Line
import geometry.Point
import kotlin.math.sqrt

class Trapezoid(points: Array<Point>) : Shape(points) {

    override fun getArea(): Double {
        val edges = this.getEdgeLines()

        val a = edges[0].length()
        val b = edges[1].length()
        val c = edges[2].length()
        val d = edges[3].length()

        val T = a - c
        val D = d * d - b * b
        val R = (T * T - D) / (2 * T)
        val h = sqrt(b * b - R * R)

        return h * (a + c) / 2
    }

    override fun isExists(): Boolean {
        val edges = this.getEdgeLines()

        val bottom = edges[0]
        val top = edges[2]
        val left = edges[1]
        val right = edges[3]

        return Line.isParallel(bottom, top) && !Line.isParallel(left, right)
    }
}
